crypto/src/lib.rs

```diff
@@ -278,58 +310,52 @@ impl From<KeyPair> for (PublicKey, PrivateKey) {
-    #[cfg_attr(not(feature="ffi_import"), debug(fmt = "{{digest: {digest_function}, payload: {}}}", "self.normalize()"))]
-    #[cfg_attr(not(feature="ffi_import"), display(fmt = "{}", "self.normalize()"))]
-    pub struct PublicKey {
-        /// Digest function
-        #[getset(get_copy = "pub")]
-        digest_function: Algorithm,
-        /// Key payload
-        payload: ConstVec<u8>,
+    #[derive(Clone, PartialEq, Eq)]
+    #[cfg_attr(not(feature="ffi_import"), derive(DeserializeFromStr, SerializeDisplay))]
+    #[cfg_attr(all(feature = "ffi_export", not(feature = "ffi_import")), ffi_type(opaque))]
+    #[allow(missing_docs)]
+    pub enum PublicKey {
+        Ed25519(ed25519::PublicKey),
+        Secp256k1(secp256k1::PublicKey),
+        BlsNormal(bls::BlsNormalPublicKey),
+        BlsSmall(Box<bls::BlsSmallPublicKey>),
     }
 }
 
 #[ffi_impl_opaque]
 impl PublicKey {
     /// Creates a new public key from raw bytes received from elsewhere
-    pub fn from_raw(algorithm: Algorithm, payload: ConstVec<u8>) -> Self {
-        Self {
-            digest_function: algorithm,
-            payload,
+    pub fn from_raw(algorithm: Algorithm, payload: &[u8]) -> Result<Self, Error> {
+        match algorithm {
+            Algorithm::Ed25519 => {
+                ed25519::Ed25519Sha512::parse_public_key(payload).map(Self::Ed25519)
+            }
+            Algorithm::Secp256k1 => {
+                secp256k1::EcdsaSecp256k1Sha256::parse_public_key(payload).map(Self::Secp256k1)
+            }
+            Algorithm::BlsNormal => bls::BlsNormal::parse_public_key(payload).map(Self::BlsNormal),
+            Algorithm::BlsSmall => bls::BlsSmall::parse_public_key(payload)
+                .map(Box::new)
+                .map(Self::BlsSmall),
         }
     }
 
     /// Extracts the raw bytes from public key
     pub fn into_raw(self) -> (Algorithm, ConstVec<u8>) {
-        (self.digest_function, self.payload)
+        (self.digest_function(), self.payload())
     }
 
     /// Key payload
-    // TODO: Derive with getset once FFI impl is fixed
-    pub fn payload(&self) -> &[u8] {
-        self.payload.as_ref()
-    }
-
-    #[cfg(feature = "std")]
-    fn try_from_private(private_key: PrivateKey) -> Result<PublicKey, Error> {
-        let digest_function = private_key.digest_function();
-        let key_gen_option = Some(KeyGenOption::FromPrivateKey(private_key));
-
-        let (public_key, _) = match digest_function {
-            Algorithm::Ed25519 => signature::ed25519::Ed25519Sha512::keypair(key_gen_option),
-            Algorithm::Secp256k1 => {
-                signature::secp256k1::EcdsaSecp256k1Sha256::keypair(key_gen_option)
-            }
-            Algorithm::BlsNormal => signature::bls::BlsNormal::keypair(key_gen_option),
-            Algorithm::BlsSmall => signature::bls::BlsSmall::keypair(key_gen_option),
-        }?;
-
-        Ok(public_key)
+    pub fn payload(&self) -> ConstVec<u8> {
```

Comment:
```suggestion
    pub fn payload(&self) -> &[u8] {
```

is this not possible? otherwise we don't need this method, `into_raw` is sufficient

Reply:
No, not possible because of temporary value.

I've found 14 references for `PublicKey::payload()` and 4 references for `PrivateKey::payload()`.

Do you think it would be more readable to use `key.into_raw().1` in 18 places rather than just `key.payload()`?

@DCNick3, what do you think?

Reply:
I created the `into_raw` function with the intention that it would promote looking at both the digest function and the payload with tuple deconstruction (`let (digest_function, payload) = key.into_raw();`). Just never got around to make the other code actually do that..

Briefly looking at the 14 references my IDE finds for `PublicKey`:
- The uses in trait implementations in `iroha_crypto` are all using the `.digest_function()`, except `Encode::size_hint`, which is not that bad
- Code in `iroha_data_model` and `iroha_p2p` that hashes the public keys IMO should really also hash the digest functions. Currently it just ignores the tag, which is fishy and what such an API can actually catch

For `PrivateKey` all the uses are in the trait implementations that also look at `digest_function` except for `Display` (maybe it actually should look at it, but this impl should probably just not exist for private keys in accordance with #3240).

With this in mind, I think it's a good idea to remove the `payload` function, as it just promotes a use that is logically incorrect (getting the key payload while ignoring the crypto algorithm it is actually used with)

P.S. we should probably also stop calling that field a digest function, because it is not. It is digital signature algorithm

Reply:
Done